packages/teams-js/src/private/externalAppAuthentication.ts

```diff
@@ -373,28 +373,55 @@ export type InvokeErrorWrapper = InvokeError & { responseType: undefined };
  */
 function validateOriginalRequestInfo(originalRequestInfo: IOriginalRequestInfo): void {
   if (originalRequestInfo.requestType === OriginalRequestType.ActionExecuteInvokeRequest) {
-    const actionExecuteRequest = originalRequestInfo as IActionExecuteInvokeRequest;
-    if (actionExecuteRequest.type !== ActionExecuteInvokeRequestType) {
-      const error: InvokeError = {
-        errorCode: InvokeErrorCode.INTERNAL_ERROR,
-        message: `Invalid action type ${actionExecuteRequest.type}. Action type must be "${ActionExecuteInvokeRequestType}"`,
-      };
-      throw error;
-    }
+    validateActionExecuteInvokeRequest(originalRequestInfo as IActionExecuteInvokeRequest);
   } else if (originalRequestInfo.requestType === OriginalRequestType.QueryMessageExtensionRequest) {
-    if (originalRequestInfo.commandId.length > 64) {
-      throw new Error('originalRequestInfo.commandId exceeds the maximum size of 64 characters');
-    }
-    if (originalRequestInfo.parameters.length > 5) {
-      throw new Error('originalRequestInfo.parameters exceeds the maximum size of 5');
+    validateQueryMessageExtensionRequest(originalRequestInfo as IQueryMessageExtensionRequest);
+  }
+}
+
+/**
+ * @beta
+ * @hidden
+ * @internal
+ * Limited to Microsoft-internal use
+ */
+export function validateActionExecuteInvokeRequest(actionExecuteRequest: IActionExecuteInvokeRequest): void {
+  if (actionExecuteRequest.type !== ActionExecuteInvokeRequestType) {
+    const error: InvokeError = {
+      errorCode: InvokeErrorCode.INTERNAL_ERROR,
+      message: `Invalid action type ${actionExecuteRequest.type}. Action type must be "${ActionExecuteInvokeRequestType}"`,
+    };
+    throw error;
+  }
+
+  if (!isPrimitiveOrPlainObject(actionExecuteRequest.data)) {
+    const error: InvokeError = {
+      errorCode: InvokeErrorCode.INTERNAL_ERROR,
+      message: `Invalid data type ${typeof actionExecuteRequest.data}. Data must be a primitive or a plain object.`,
+    };
+    throw error;
+  }
+}
+
+/**
+ * @beta
+ * @hidden
+ * @internal
+ * Limited to Microsoft-internal use
+ */
+function validateQueryMessageExtensionRequest(originalRequestInfo: IQueryMessageExtensionRequest): void {
+  if (originalRequestInfo.commandId.length > 64) {
```

Comment:
I might not know much of the background context, and I feel confused and curious about where these magic numbers (like 64, 5, 512) come from.

Reply:
They all came from 3pcxp team. In this PR, I am just dividing the already existing function into smaller functions. These numbers were added when the support was added for MEs in last December